cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -1,55 +1,247 @@
 mod cache;
-mod transaction_pattern;
 mod web3_connector;
 
-pub use self::{
-    cache::Cache,
-    transaction_pattern::{Event, Topic, TransactionPattern, TRANSACTION_STATUS_OK},
-    web3_connector::Web3Connector,
-};
+pub use self::{cache::Cache, web3_connector::Web3Connector};
 use crate::{
     btsieve::{BlockByHash, LatestBlock, Predates, ReceiptByHash},
-    ethereum::{Transaction, TransactionAndReceipt, TransactionReceipt, H256, U256},
+    ethereum::{Address, Bytes, Transaction, TransactionReceipt, H256, U256},
     Never,
 };
 use anyhow;
 use chrono::NaiveDateTime;
-use futures_core::compat::Future01CompatExt;
+use ethbloom::Input;
+use futures_core::{compat::Future01CompatExt, Future};
 use genawaiter::{
     sync::{Co, Gen},
     GeneratorState,
 };
-use std::collections::HashSet;
+use std::{collections::HashSet, pin::Pin};
 
 type Hash = H256;
 type Block = crate::ethereum::Block<Transaction>;
 
-pub async fn matching_transaction<C>(
+pub const TRANSACTION_STATUS_OK: u32 = 1;
+
+pub async fn matching_create_contract_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    bytecode: Bytes,
+) -> anyhow::Result<(Transaction, Address)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    let (transaction, receipt) =
+        matching_transaction(blockchain_connector, start_of_swap, |transaction| {
+            // transaction.to address is None if, and only if, the transaction
+            // creates a contract.
+            transaction.to.is_none() && transaction.input == bytecode
+        })
+        .await?;
+
+    match receipt.contract_address {
+        Some(location) => Ok((transaction, location)),
+        None => return Err(anyhow::anyhow!("contract address missing from receipt")),
+    }
+}
+
+pub async fn matching_event_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    event: Event,
+    action: &'static str,
+) -> anyhow::Result<(Transaction, TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction_bloom_filter(
+        blockchain_connector.clone(),
```

Comment:
Similarly, with the matcher being non-async, you should be able to avoid the clone because you don't need the connector out here.

Reply:
This ones right, rustc caught this one.  On a separate note, I would be interested in going through the code base and trying to remove clones at some stage when we are playing around - bit early for optimisations I know but this would be a good way to learn more about Rust life times, which is something I'd like to learn more about.

Reply:
Cloning things per se is not bad, even though the Rust community is religious about it :sweat_smile: 

What is important is thinking about _what_ you are cloning here. The connector is not data, it is functionality that we are passing around. Maybe we should make that explicit and always take an `Arc<C>` to make it totally clear that cloning it is:

a) cheap
b) not harmful in any way

Reply:
> @tcharding do you plan on doing the bitcoin side also in this PR?
> If no, please open an issue for it so we can track the remaining work!

I have not reached a decision yet, I was planning on finishing the Ethereum side then once I know exactly whats involved looking into the Bitcoin code and seeing if we should do this now or if we should make the Bitcoin code use Generators etc (all the improvements that have been done only on the Ethereum side).  Open to suggestions, will definitely create an issue if we go that way.

Reply:
There is already a ticket for adopting the use of generators (https://github.com/comit-network/comit-rs/issues/1878). Happy to do that one first and then follow-up with adopting the model proposed in this PR for the Bitcoin side.